### PR TITLE
oxen: init at 9.1.3

### PR DESCRIPTION
**Pull request by viric**
Based on Monero expression

###### Motivation for this change

Add the oxen wallet

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by viric** [comment]
Oops pslab doesn't belong here.

File: pkgs/tools/networking/unbound/default.nix [hunk 1]
```diff
@@ -37,7 +37,8 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [ makeWrapper ];
 
-  buildInputs = [ openssl nettle expat libevent ]
+  propagatedBuildInputs = [ libevent ];
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/unbound/default.nix:40]: You are sure about this change?

File: pkgs/applications/blockchains/oxen/default.nix [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  src = fetchFromGitHub {
+    owner = "oxen-io";
+    repo = "oxen-core";
+    rev = "v${version}";
+    sha256 = "11g3wqn0syk47yfcsdql5737kpad8skwdxhifn2yaz9zy8n3xqqb";
+    fetchSubmodules = true;
+  };
+
+  # Required for static linking, the only supported install path
+  lbzmqsrc = fetchurl {
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/oxen/default.nix:30]: Maybe we should move this to an extra package that it receives update etc.

File: pkgs/applications/blockchains/oxen/default.nix [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+stdenv.mkDerivation rec {
+  pname = "oxen";
+  version = "9.1.3";
+
+  src = fetchFromGitHub {
+    owner = "oxen-io";
+    repo = "oxen-core";
+    rev = "v${version}";
+    sha256 = "11g3wqn0syk47yfcsdql5737kpad8skwdxhifn2yaz9zy8n3xqqb";
+    fetchSubmodules = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/oxen/default.nix:26]: There are some more vendored libraries we can remove.

Comment (viric) [reply pkgs/applications/blockchains/oxen/default.nix:26]: Some are forks. And others are difficult to take out. I already tried a bit.

File: pkgs/applications/blockchains/oxen/default.nix (hunk 2 above)
Comment (viric) [reply pkgs/applications/blockchains/oxen/default.nix:30]: It has to match exactly the bithash that they will use in their particular receipt for static build. I'd leave it here. They know how they want it.

File: pkgs/tools/networking/unbound/default.nix (hunk 1 above)
Comment (viric) [reply pkgs/tools/networking/unbound/default.nix:40]: The unbound pkgconfig says:
Requires.private:   libcrypto libssl libevent

Maybe its pkgconfig is wrong (meant for static lib). Likely I already had openssl and I missed simply libevent here, so I added it.
Should I add openssl to propagated? I don't want to patch its pkgconfig.

File: pkgs/applications/blockchains/oxen/default.nix [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+, trezorSupport ? true
+,   libusb1  ? null
+,   protobuf ? null
+,   python3  ? null
+}:
+
+assert stdenv.isDarwin -> IOKit != null;
+assert trezorSupport -> lib.all (x: x!=null) [ libusb1 protobuf python3 ];
+
+stdenv.mkDerivation rec {
```
[the remark this answers is not in the document]
Comment (viric) [reply pkgs/applications/blockchains/oxen/default.nix:17]: I cannot test Darwin, so maybe I rather remove all these things that simply come from the monero template I took.

File: pkgs/applications/blockchains/oxen/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/applications/blockchains/oxen/default.nix:26]: Thats a pity.

File: pkgs/applications/blockchains/oxen/default.nix [hunk 5]
```diff
@@ -0,0 +1,67 @@
+{ stdenv, lib, fetchurl, fetchFromGitHub, fetchpatch
+, cmake, pkg-config
+, boost, openssl, unbound
+, pcsclite, readline, libsodium, hidapi
+, rapidjson
+, curl, sqlite
+, trezorSupport ? true
+,   libusb1  ? null
+,   protobuf ? null
+,   python3  ? null
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/oxen/default.nix:10]:
```suggestion
, trezorSupport ? true
, libusb1
, protobuf
, python3
```
They cannot be null. If they are null there is a mistake here.

File: pkgs/applications/blockchains/oxen/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/applications/blockchains/oxen/default.nix:22]: Hmmm, well, not great but nothing we can change.

File: pkgs/applications/blockchains/oxen/default.nix [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    protobuf curl sqlite
+  ] ++ lib.optionals trezorSupport [ libusb1 protobuf python3 ];
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
+    # "-DUSE_DEVICE_TREZOR=ON"
+    # "-DBUILD_GUI_DEPS=ON"
+    "-DReadline_ROOT_DIR=${readline.dev}"
+    # It build with shared libs but doesn't install them. Fail.
+    # "-DBUILD_SHARED_LIBS=ON"
```
Comment (jonringer) [inline pkgs/applications/blockchains/oxen/default.nix:56]: if the binaries still link against them, they should be installed

File: pkgs/tools/networking/unbound/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/tools/networking/unbound/default.nix:40]: for dynamically linked programs, the `.private` libs should not be propagated. Please put libevent back into `buildInputs`

File: pkgs/applications/blockchains/oxen/default.nix (hunk 6 above)
Comment (viric) [reply pkgs/applications/blockchains/oxen/default.nix:56]: The static build doesn't use shared libs so all works fine.

File: pkgs/tools/networking/unbound/default.nix (hunk 1 above)
Comment (viric) [reply pkgs/tools/networking/unbound/default.nix:40]: Aah right. I suspected that. In fact today I have no idea what was that change about... I prepared the change long before the PR for my own use and likely it's completely unrelated to oxen. I'll take it out!

File: pkgs/applications/blockchains/oxen/default.nix [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  postPatch = ''
+    # remove vendored libraries
+    rm -r external/rapidjson
+    sed -i s,/lib/,/lib64/, external/loki-mq/cmake/local-libzmq//LocalLibzmq.cmake
+  '';
+
+  postInstall = ''
+    rm -R $out/lib $out/include
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/oxen/default.nix:38]:
```suggestion
    rm -r $out/lib $out/include
```

**Pull request merged**